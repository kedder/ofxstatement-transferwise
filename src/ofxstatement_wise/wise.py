from csv import DictReader
from datetime import datetime
from typing import Dict, Iterable, Optional
from decimal import Decimal

from ofxstatement.plugin import Plugin
from ofxstatement.parser import StatementParser
from ofxstatement.statement import (
    Statement,
    StatementLine,
    BankAccount,
)


class TransferwisePlugin(Plugin):
    """Transferwise CSV format"""

    def get_parser(self, filename: str) -> "TransferwiseParser":
        default_ccy = self.settings.get("currency")
        account_id = self.settings.get("account")
        return TransferwiseParser(filename, default_ccy, account_id)


class TransferwiseParser(StatementParser[Dict[str, str]]):
    def __init__(
        self, filename: str, currency: str | None = None, account_id: str | None = None
    ) -> None:
        super().__init__()
        self.filename = filename
        self.currency = currency
        self.account_id = account_id

    def parse(self) -> Statement:
        stmt = super().parse()
        stmt.currency = self.currency
        stmt.account_id = self.account_id
        return stmt

    def split_records(self) -> Iterable[Dict[str, str]]:
        with open(self.filename, "rt") as f:
            yield from DictReader(f)

    def parse_record(self, line: Dict[str, str]) -> Optional[StatementLine]:
        """Parse given transaction line and return StatementLine object"""
        sl = StatementLine()

        sl.id = line["TransferWise ID"]
        sl.date = datetime.strptime(line["Date Time"], "%d-%m-%Y %H:%M:%S.%f")
        sl.memo = line["Description"]
        sl.amount = Decimal(line["Amount"])

        currency = line["Currency"]
        if currency != self.currency:
            # Skip lines in some other currencies
            return None

        sl.memo = self._make_memo(line)

        payee_acc_no = line["Payee Account Number"]
        if payee_acc_no:
            sl.bank_account_to = BankAccount("", payee_acc_no)

        assert sl.amount is not None
        sl.trntype = line["Transaction Type"]
        return sl

    def _make_memo(self, line: Dict[str, str]) -> str:
        descr = line["Description"]
        payref = line["Payment Reference"]
        exc_from = line["Exchange From"]
        exc_to = line["Exchange To"]
        exc_rate = line["Exchange Rate"]

        memo = descr
        if payref:
            memo += f" ({payref})"
        if exc_from and exc_to and exc_rate:
            memo += f", {exc_rate} {exc_from}/{exc_to}"
        return memo
