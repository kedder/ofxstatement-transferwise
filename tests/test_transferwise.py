import os

from ofxstatement.ui import UI

from ofxstatement_wise.wise import TransferwisePlugin


def test_transferwise() -> None:
    config = {"currency": "USD", "account": "TW1"}
    plugin = TransferwisePlugin(UI(), config)
    here = os.path.dirname(__file__)
    sample_filename = os.path.join(here, "sample-statement.csv")

    parser = plugin.get_parser(sample_filename)
    statement = parser.parse()

    assert statement is not None

    assert len(statement.lines) == 5
    # all ids are unique
    assert len(set(ln.id for ln in statement.lines)) == 5
    assert all(ln.amount for ln in statement.lines)
