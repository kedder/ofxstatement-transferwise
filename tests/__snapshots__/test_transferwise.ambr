# serializer version: 1
# name: test_transferwise
  <Statement> {   'account_id': 'TW1',
      'account_type': 'CHECKING',
      'bank_id': None,
      'currency': 'USD',
      'invest_lines': [],
      'lines': [   <StatementLine> {   'amount': Decimal('-357.38'),
      'bank_account_to': <BankAccount> {   'acct_id': 'LT21 1111 2222 3333 4444',
      'acct_key': None,
      'acct_type': 'CHECKING',
      'bank_id': '',
      'branch_id': None},
      'check_no': None,
      'date': datetime.datetime(2020, 8, 24, 0, 0),
      'date_user': None,
      'id': 'be6408d9bbe3e3911894f376ea61db777b6ad6b3',
      'memo': 'Sent money to John Doe (Moving to Revolut), 0.84840 USD/EUR',
      'payee': None,
      'refnum': 'TRANSFER-175545673',
      'trntype': 'CREDIT'},
                   <StatementLine> {   'amount': Decimal('-11.35'),
      'check_no': None,
      'date': datetime.datetime(2020, 8, 24, 0, 0),
      'date_user': None,
      'id': '5c4f8d3a9cf52c60b5f10ebfad25103e129326f7',
      'memo': 'TransferWise Charges for: TRANSFER-175545673 (Sending money)',
      'payee': None,
      'refnum': 'TRANSFER-175545673',
      'trntype': 'CREDIT'},
                   <StatementLine> {   'amount': Decimal('125.00'),
      'check_no': None,
      'date': datetime.datetime(2020, 8, 24, 0, 0),
      'date_user': None,
      'id': 'a474ecf34286b851894500d6ec1d73e94a795a10',
      'memo': 'Received money from the friend. with reference ',
      'payee': None,
      'refnum': 'TRANSFER-175522746',
      'trntype': 'DEBIT'},
                   <StatementLine> {   'amount': Decimal('20.17'),
      'check_no': None,
      'date': datetime.datetime(2020, 8, 20, 0, 0),
      'date_user': None,
      'id': '99554757a35ed7163bd936f08ff83f17c0c73921',
      'memo': 'Topped up balance',
      'payee': None,
      'refnum': 'TRANSFER-174549828',
      'trntype': 'DEBIT'},
                   <StatementLine> {   'amount': Decimal('-0.17'),
      'check_no': None,
      'date': datetime.datetime(2020, 8, 20, 0, 0),
      'date_user': None,
      'id': '3f6edbaa1763c02fb5c5a702a5a7697bb3d4b881',
      'memo': 'TransferWise Charges for: TRANSFER-174549828',
      'payee': None,
      'refnum': 'TRANSFER-174549828',
      'trntype': 'CREDIT'}]}
# ---
